File: 234. Palindrome Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

ListNode* reverseLinkList(ListNode* node) { 

    ListNode* head = NULL;
    while( node ) { 
        ListNode* curNode = new ListNode(node->val);
        if( head ) { 
            curNode->next = head;
            head = curNode;
        } else head = curNode;
        node = node->next;

    }
    return head;
}

bool checkPalindrome(ListNode* front, ListNode* back) { 
    ListNode* fast = front;
    while( fast  ) { 

        if( front->val != back->val ) return false;
        front = front->next;
        back = back->next;
        fast = fast->next ? fast->next->next : fast->next;
    }
   
    return true;

}
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode* reverseHead = reverseLinkList(head);
        return checkPalindrome(head, reverseHead);
    }
};

File: python/Lakshay/234. Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

def reverseList(head: Optional[ListNode]) -> Optional[ListNode]:
        root = None
        prv = None
        while head:
            curNode = ListNode(head.val)
            root = curNode
            root.next = prv
            prv = curNode
            head = head.next
        
        return root
class Solution:
    
    
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        
        reverse_list = reverseList(head)
        while head and reverse_list:
            if head.val != reverse_list.val:
                return False
            head = head.next
            reverse_list = reverse_list.next
        
        if head or reverse_list:
            return False
        
        return True
